File: kotlin/kotlin-demo/src/main/kotlin/tech/techstay/oop/Class.kt
package tech.techstay.oop

class Empty
data class Person(val name: String, val age: Int) {
    val nickname
        get() = this.name.substring(0..3)

    constructor() : this("default name", 0)
}

class DontCreateMe private constructor() {
}

fun main() {

    // can't be created
//    val dontCreateMe = DontCreateMe()
    val person = Person("techstay", 18)
    println(person.nickname)
}

File: scala/scala-example/src/main/scala/tech/techstay/oop/Class.scala
package tech.techstay.oop

class Person(var id: Int, var name: String, var age: Int):
  // constructors
  def this() =
    this(0, "", 0)

  def selfIntroduce(): Unit =
    println(s"I'm $name")

  override def toString: String =
    s"Person(id:$id, name:$name, age:$age)"

end Person

@main def clazz() =
  val p1 = Person(1, "techstay", 18)
  println(p1)
  // accessing public fields
  p1.name = "TECHSTAY"
  println(p1)
  p1.selfIntroduce()
